sdv/multi_table/hma.py

```diff
@@ -216,3 +216,309 @@ def _fit(self, processed_data):
+
+        Args:
+            sampled_data (dict):
+                Generated data
+
+        Return:
+            pandas.DataFrame:
+                Formatted synthesized data.
+        """
+        final_data = {}
+        for table_name, table_rows in sampled_data.items():
+            parents = self.metadata._get_parent_map().get(table_name)
+            if parents:
+                for parent_name in parents:
+                    foreign_keys = self._get_foreign_keys(parent_name, table_name)
+                    for foreign_key in foreign_keys:
+                        if foreign_key not in table_rows:
+                            parent_ids = self._find_parent_ids(
+                                table_name,
+                                parent_name,
+                                foreign_key,
+                                sampled_data
+                            )
+                            table_rows[foreign_key] = parent_ids
+
+            synthesizer = self._table_synthesizers.get(table_name)
+            dtypes = synthesizer._data_processor._dtypes
+            for name, dtype in dtypes.items():
+                table_rows[name] = table_rows[name].dropna().astype(dtype)
+
+            final_data[table_name] = table_rows[list(dtypes.keys())]
+
+        return final_data
+
+    def _extract_parameters(self, parent_row, table_name, foreign_key):
+        """Get the params from a generated parent row.
+
+        Args:
+            parent_row (pandas.Series):
+                A generated parent row.
+            table_name (str):
+                Name of the table to make the model for.
+            foreign_key (str):
+                Name of the foreign key used to form this
+                parent child relationship.
+        """
+        prefix = f'__{table_name}__{foreign_key}__'
+        keys = [key for key in parent_row.keys() if key.startswith(prefix)]
+        new_keys = {key: key[len(prefix):] for key in keys}
+        flat_parameters = parent_row[keys].fillna(0)
+
+        num_rows_key = f'{prefix}num_rows'
+        if num_rows_key in flat_parameters:
+            num_rows = flat_parameters[num_rows_key]
+            flat_parameters[num_rows_key] = min(self._max_child_rows[num_rows_key], num_rows)
+
+        return flat_parameters.rename(new_keys).to_dict()
+
+    def _sample_rows(self, model, table_name, num_rows=None):
+        """Sample ``num_rows`` from ``model``.
+
+        Args:
+            model (copula.multivariate.base):
+                Fitted model.
+            table_name (str):
+                Name of the table to sample from.
+            num_rows (int):
+                Number of rows to sample.
+
+        Returns:
+            pandas.DataFrame:
+                Sampled rows, shape (, num_rows)
+        """
+        num_rows = num_rows or model._num_rows
+        sampled_rows = model._sample(num_rows)
+
+        data_processor = self._table_synthesizers[table_name]._data_processor
```

Comment:
Maybe we just need to store the metadata and data processor for each table instead of the whole synthesizer since we actually recreate the model part. Or conversely, maybe we update the underlying model on the stored synthesizer and then just call `sample`